compiler/src/dmd/identifier.d

```diff
@@ -300,6 +300,33 @@ nothrow:
         sv.value = id;
     }
 
+    extern (D) static Identifier idPoolNormalize(scope const(char)[] s, bool isNormalized)
+    {
+        // Normalization itself is not implemented here currently.
+
+        return idPool(s, false);
+
+        /*
+        Pseudo code for how this function needs to work:
+
+        if (s in stringtable)
+            return stringtable[s];
+        else if (isNormalized)
+            return idPool(s, false);
+
+        auto sNormalized = toNFC(s);
+        if (sNormalized in stringtable)
+        {
+            free(sNormalized);
```

Comment:
Mem.xfree?

Reply:
The exact function doesn't matter it's pseudo code.

What I'm worried about is whether the string interner ``stringtable`` is capable of handling the pattern of storing unnormalized as key to normalized value.

Write an identifier one way 100 times, which should result in normalization only the 1 time if it wasn't normalized.

This is something I need feedback on before I start next PR in a few months time.

Reply:
stringtable uses a hash of the bytes in the identifier. It does not attempt normalization. Doing that would slow it way down.

Reply:
Yes, that isn't what I am suggesting.

The stringtable is only responsible for storing key to value map. Unlike currently I would need to make the key be a different string to the value.

Turns out this is as simple as adding an extra parameter (key) which gets sent to the hash instead of the value parameter.

The quick check algorithm quickly determines if normalization needs to occur for Unicode, for ASCII its always off.

``idPoolNormalize`` then takes whether it is currently normalized and performs the normalization calling into the stringtable as describe in the pseudo code.

I do believe this is all solved for next PR.